frontend/src/main/scala/bloop/testing/TestInternals.scala

```diff
@@ -29,14 +37,83 @@ object TestInternals {
     new FilteredLoader(getClass.getClassLoader, filter)
   }
 
-  @tailrec
-  def executeTasks(tasks: List[TestTask], eventHandler: EventHandler, logger: Logger): Unit = {
-    tasks match {
-      case task :: rest =>
-        val newTasks = task.execute(eventHandler, Array(logger)).toList
-        executeTasks(rest ::: newTasks, eventHandler, logger)
-      case Nil =>
-        ()
+  /**
+   * Execute the test tasks.
+   *
+   * @param processConfig   Configures whether tests should be run in a forked JVM or in process.
+   * @param discoveredTests The tests that were discovered.
+   * @param eventHandler    Handler that reacts on messages from the testing frameworks.
+   * @param logger          Logger receiving test output.
+   */
+  def executeTasks(processConfig: ProcessConfig,
+                   discoveredTests: DiscoveredTests,
+                   eventHandler: EventHandler,
+                   logger: Logger): Unit = {
+    processConfig match {
+      case inProcess: InProcess => executeTasks(inProcess, discoveredTests, eventHandler, logger)
+      case fork: Fork => executeTasks(fork, discoveredTests, eventHandler, logger)
+    }
+  }
+
+  /**
+   * Execute the test tasks in a forked JVM.
+   *
+   * @param fork            Configuration for the forked JVM.
+   * @param discoveredTests The tests that were discovered.
+   * @param eventHandler    Handler that reacts on messages from the testing frameworks.
+   * @param logger          Logger receiving test output.
+   */
+  private def executeTasks(fork: Fork,
+                           discoveredTests: DiscoveredTests,
+                           eventHandler: EventHandler,
+                           logger: Logger): Unit = {
+    logger.debug("Starting forked test execution.")
+
+    val testLoader = fork.toClassLoader(Some(filteredLoader))
+    val server = new TestServer(logger, eventHandler, discoveredTests)
+    val forkMain = classOf[sbt.ForkMain].getCanonicalName
+    val arguments = Array(server.port.toString)
+    val testAgentFiles = DependencyResolution.resolve(sbtOrg, testAgentId, testAgentVersion, logger)
```

Comment:
Oh wow... is it possible that we don't resolve in such an important part of the pipeline? Can we not first try to locate and reuse the test agent jar from our current classloader by checking the JDK protection domain (because i think we already have it in bloop's classloader), and only if that fails we try to resolve? I think this is too heavy here, even if coursier is being used.

Reply:
I'll benchmark that to see if it's really that expensive.